datafusion/expr/src/udf.rs

```diff
@@ -17,12 +17,67 @@
 
 //! Udf module contains foundational types that are used to represent UDFs in DataFusion.
 
-use crate::{Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature};
+use crate::{
+    ColumnarValue, Expr, FuncMonotonicity, ReturnTypeFunction,
+    ScalarFunctionImplementation, Signature, TypeSignature, Volatility,
+};
+use arrow::array::ArrayRef;
+use arrow::datatypes::DataType;
+use datafusion_common::{internal_err, DataFusionError, Result};
+use std::any::Any;
 use std::fmt;
 use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
+// TODO(PR): add doc comments
+pub trait ScalarFunctionDef: Any + Sync + Send + std::fmt::Debug {
+    /// Return as [`Any`] so that it can be
+    /// downcast to a specific implementation.
+    fn as_any(&self) -> &dyn Any;
+
+    // May return 1 or more name as aliasing
+    fn name(&self) -> &[&str];
+
+    fn input_type(&self) -> TypeSignature;
+
+    fn return_type(&self) -> FunctionReturnType;
+
+    fn execute(&self, _args: &[ArrayRef]) -> Result<ArrayRef> {
+        internal_err!("This method should be implemented if `supports_execute_raw()` returns `false`")
+    }
+
+    fn volatility(&self) -> Volatility;
+
+    fn monotonicity(&self) -> Option<FuncMonotonicity>;
+
+    // ===============================
+    // OPTIONAL METHODS START BELOW
```

Comment:
This trait consists of mandatory and optional methods, it can get a bit lengthy...
An alternative implementation is trait inheritance(e.g. `pub trait ScalarFunctionExtended: ScalarFunctionDef`, it seems won't be much clearer than the current one.
We can add more docs/examples to make it more straightforward later